relayer/src/link/relay_path.rs

```diff
@@ -693,103 +733,212 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+            .timestamp();
+        let duration = Timestamp::now()
+            .duration_since(&chain_time)
+            .unwrap_or_default();
+        Ok(Instant::now().sub(duration))
     }
 
     /// Handles updating the client on the destination chain
+    /// Returns the height at which the client update was processed
     fn update_client_dst(
         &self,
         src_chain_height: Height,
         tracking_id: &str,
-    ) -> Result<(), LinkError> {
-        // Handle the update on the destination chain
-        // Check if a consensus state at update_height exists on destination chain already
-        if self
-            .dst_chain()
-            .proven_client_consensus(self.dst_client_id(), src_chain_height, Height::zero())
-            .is_ok()
-        {
-            return Ok(());
-        }
-
-        let mut dst_err_ev = None;
-        for i in 0..MAX_RETRIES {
-            let dst_update = self.build_update_client_on_dst(src_chain_height)?;
-            info!(
-                "sending updateClient to client hosted on destination chain for height {} [try {}/{}]",
-                src_chain_height,
-                i + 1, MAX_RETRIES,
-            );
-
-            let tm = TrackedMsgs::new(dst_update, tracking_id);
-
-            let dst_tx_events = self
-                .dst_chain()
-                .send_messages_and_wait_commit(tm)
-                .map_err(LinkError::relayer)?;
-            info!("result: {}", PrettyEvents(&dst_tx_events));
+    ) -> Result<Height, LinkError> {
+        self.do_update_client_dst(src_chain_height, tracking_id, MAX_RETRIES)
+    }
 
-            dst_err_ev = dst_tx_events
-                .into_iter()
-                .find(|event| matches!(event, IbcEvent::ChainError(_)));
+    /// Perform actual update_client_dst with retries.
+    ///
+    /// Note that the retry is only performed in the case when there
+    /// is a ChainError event. It would return error immediately if
+    /// there are other errors returned from calls such as
+    /// build_update_client_on_dst.
+    fn do_update_client_dst(
+        &self,
+        src_chain_height: Height,
+        tracking_id: &str,
+        retries_left: usize,
+    ) -> Result<Height, LinkError> {
+        info!( "sending update_client to client hosted on source chain for height {} (retries left: {})", src_chain_height, retries_left );
 
-            if dst_err_ev.is_none() {
-                return Ok(());
+        let dst_update = self.build_update_client_on_dst(src_chain_height)?;
+        let tm = TrackedMsgs::new(dst_update, tracking_id);
+        let dst_tx_events = self
+            .dst_chain()
+            .send_messages_and_wait_commit(tm)
+            .map_err(LinkError::relayer)?;
+        info!("result: {}", PrettyEvents(&dst_tx_events));
+
+        let (error, update, other) = Self::event_per_type(dst_tx_events);
+        match (error, update, other) {
+            (None, Some(update_event), None) => Ok(update_event.height()),
```

Comment:
Would be helpful to write out detailed comments on what it really means when each of these branches are triggered.

Reply:
https://github.com/informalsystems/ibc-rs/pull/1917/commits/14dcb2ae44e47c17a0b03a56aa32d1a74840ee3e